src/position-observer.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {SubscriptionApi} from './iframe-helper';
+import {MessageType} from './3p-frame-messaging.js';
+import {rateLimit} from './utils/rate-limit';
+import {LayoutRectDef} from './layout-rect';
+
+/** @const */
+const MIN_EVENT_INTERVAL_IN_MS = 100;
+
+/**
+ * @typedef {{
+ *   viewport: !LayoutRectDef,
+ *   target: !LayoutRectDef
+ * }}
+ */
+let PositionEntryDef;
+
+export class PositionObserverApi {
+  /**
+   * @param {!AMP.BaseElement} baseElement
+   * @param {!Element} iframe
+   * @param {boolean=} opt_is3p
+   */
+  constructor(baseElement, iframe, opt_is3p) {
+    /** @private @const {!AMP.BaseElement} */
+    this.baseElement_ = baseElement;
```

Comment:
Ideally this should be any element. However that would require me to call `getBoundingClientRect()` which is expensive I think. Given that it is an AMP element, I can get viewport rect and element layoutBox with no extra cost. Also AMP element can deal with current use cases, I would like to make it AMP element unless there're further requirements in the future.